test/test_optimizers.f90

```diff
@@ -0,0 +1,150 @@
+program test_optimizers
+  use nf, only: dense, input, network
+  use nf_loss, only: quadratic
+  use iso_fortran_env, only: stderr => error_unit
+  use nf_optimizers, only: sgd, rmsprop
+  implicit none
+  type(network) :: net(5)
+  real, allocatable :: x(:), y(:)
+  real, allocatable :: ypred(:)
+  integer, parameter :: num_iterations = 1000
+  real :: tolerance = 1e-3
+  integer :: n, i
+  logical :: ok_sgd = .true., ok_momentum = .true., ok_nesterov = .true., ok_rmsprop = .true.
+  logical :: converged = .false.
+  ! Initialize loss and prev_loss
+  real :: loss, prev_loss
+  prev_loss = 0.0
```

Comment:
These are used in the `check_convergence` function; it's generally less readable and more error-prone to keep the global state like this. Declare `loss` inside the function.